apacheconfig/wloader.py

```diff
@@ -0,0 +1,204 @@
+        """
+
+
+class ItemNode(Node):
+    """Creates object containing data for a comment or option-value directive.
+
+    Also manages any preceding whitespace. Can represent a key/value option,
+    a comment, or an include/includeoptional directive.
+
+    Examples of what ItemNode fields might look like for different directives::
+
+        "option"
+            name: "option", value: None, whitespace: ""
+        "include relative/path/*"
+            name: "include", value: "relative/path/*", whitespace: ""
+        "\\n  option = value"
+            name: "option", value: "value", whitespace: "\\n  "
+        "# here is a comment"
+            name: "# here is a comment", value: None, whitespace: ""
+        "\\n  # here is a comment"
+            name: "# here is a comment", value: None, whitespace: "\\n  "
+
+    To construct from a raw string, use the `parse` constructor. The regular
+    constructor receives data from the internal apacheconfig parser.
+
+    Args:
+        raw (list): Raw data returned from ``apacheconfig.parser``.
+    """
+
+    def __init__(self, raw, options={}):
+        self._type = raw[0]
+        self._raw = tuple(raw[1:])
+        self._whitespace = ""
+        if len(raw) > 1 and raw[1].isspace():
+            self._whitespace = raw[1]
+            self._raw = tuple(raw[2:])
+
+    @property
+    def ast_node_type(self):
+        """Returns object typestring as defined by the apacheconfig parser.
+
+        Returns:
+            The typestring (as defined by the apacheconfig parser) for
+            this node. The possible values for this are ``"comment"``,
+            ``"statement"``, ``"include"``, ``"includeoptional"``.
+        """
+        return self._type
+
+    @property
+    def whitespace(self):
+        """Returns preceding whitespace for this node.
+
+        For example::
+
+            ItemNode('\\n  option value').whitespace => "\\n  "
+            ItemNode('option value').whitespace => ""
+            ItemNode('\\n  # comment').whitespace => "\\n  "
+        """
+        return self._whitespace
+
+    @whitespace.setter
+    def whitespace(self, value):
+        """See base class. Operates on preceding whitespace."""
+        self._whitespace = value
+
+    @staticmethod
+    def parse(raw_str, options={}, parser=None):
+        """Factory for :class:`apacheconfig.ItemNode` by parsing data from a
+        config string.
+
+        Args:
+            options (dict): Additional options to pass to the created parser.
+                Ignored if another ``parser`` is supplied.
+            parser (:class:`apacheconfig.ApacheConfigParser`): optional, to
+                re-use an existing parser. If ``None``, creates a new one.
+        Returns:
+            :class:`apacheconfig.ItemNode` containing metadata parsed from
+            ``raw_str``.
+        """
+        if not parser:
+            parser = _create_apache_parser(options, start='startitem')
```

Comment:
I am yet to wrap my head around it, however my concern would be that creating and not reusing the entire stack might be expensive.

Reply:
Yes, the intention was for folks to create a parser and pass it in! I exposed the default apache parser factory fn and made it a required parameter for `Item.parse`.